issue #3304 - favor Exceptions over Errors in ExampleRequestProcessor

lmsurpre opened the pull request:
The method signature indicates that Exception is thrown, but the
implementation was throwing AssertionErrors instead.
This lead us to miss these in our exception handling and, since they
were within an ExecutorPool, it never got bubbled up to the test thread.

After this change, 3 examples from our MINIMAL set were failing due to 
the server defaulting to fhirVersion 4.0 and so we now explicitly set the 
fhirVersion to 4.3 in ExampleRequestProcessor.

I made some other minor changes (mostly style preference stuff) along
the way and included those in the commit.

Signed-off-by: Lee Surprenant <lmsurpre@merative.com>

punktilious commented on fhir-model/src/test/java/com/ibm/fhir/model/spec/test/R4ExamplesDriver.java:
```diff
@@ -261,17 +262,15 @@ public void processExample(String file, Format format, Expectation expectation)
                 actual = processExample(file, resource, expectation);
             }
         } catch (ExampleProcessorException e) {
+            if (firstException == null) {
```
What is the purpose of changing the logic here?

lmsurpre replied:
I wasn't sure why we weren't getting firstException set.
I realized that if processExample (line 268) threw an ExampleProcessorException then we'd never set it.
So I took the logic from the old lines 266-268 and moved them down here instead so that we apply them no matter where ExampleProcessorException is thrown from.

lmsurpre replied:
It turned out not to be our issue, but I thought the change was good and so I left it.

punktilious commented on fhir-model/src/test/java/com/ibm/fhir/model/spec/test/R4ExamplesDriver.java:
```diff
@@ -180,56 +177,60 @@ public void processExample(String file, Expectation expectation) throws ExampleP
-                catch (InterruptedException x) {
-                    // NOP
-                }
+        if (pool == null) {
```
why the switch? Now there are multiple exits from the method.

lmsurpre replied:
just a preference thing.  with the long if block i find it hard to remember the context when we get to that bottom else.  i can change it back to how it was or i could just handle this case first, remove the return, and put the rest into an else

punktilious commented:
If you're going to change an exception variable from e to x you'll be changing it in a lot of places - I use x all the time.

punktilious reviewed (approved):
LGTM

lmsurpre commented:
> If you're going to change an exception variable from e to x you'll be changing it in a lot of places - I use x all the time.

That one wasn't really purposeful.  At first I changed the catch to catch "Error" and so I changed the x to and e since x didn't make sense to me for Error.  But it does make sense for Exception (although I do generally use 'e' for that).
Anyway, I changed it back to exception but never changed the variable name back to x.  Should be fine, but don't expect a bunch of PRs changing `x` to `e` :-)
